### PR TITLE
🚧Add samples to readme

### LINKED ISSUE #92: 🚧Add samples to readme
### Complete The Item Below

- [X] I have updated the title without removing the 🚧 emoji.

### Description

Update the project README to have a couple of code samples.

### Acceptance Criteria

- [x] Features section added
- [x] Code samples added

### ToDo Items

- [X] Change type labels added to this issue.  Refer to the _**Change Type Labels**_ section below.
- [X] Priority label added to this issue.  Refer to the _**Priority Type Labels**_ section below.
- [X] Issue linked to the correct project _(if applicable)_.
- [x] Issue linked to the correct milestone _(if applicable)_.
- [x] Draft pull request created and linked to this issue _(only required with code changes)_.

### Issue Dependencies

_No response_

### Related Work

_No response_

### Additional Information:

**_<details closed><summary>Change Type Labels</summary>_**

| Change Type         | Label                |
|---------------------|----------------------|
| Bug Fixes           | `🐛bug` |
| Breaking Changes    | `🧨breaking changes` |
| New Feature         | `✨new feature` |
| CICD Changes        | `♻️cicd` |
| Code Doc Changes    | `🗒️documentation/code` |
| Product Doc Changes | `📝documentation/product` |
</details>

**_<details closed><summary>Priority Type Labels</summary>_**

| Priority Type       | Label             |
|---------------------|-------------------|
| Low Priority        | `low priority`    |
| Medium Priority     | `medium priority` |
| High Priority       | `high priority`   |
</details>

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct.

### PR DESCRIPTION
**Pull request by CalvinWilkinson**
### **Description:**

To allow this pull request to be merged, please make sure that the following items of this pull request are complete.

This pull request closes #92

✅The pull request head branch must be a feature branch with the syntax 'feature/<issue-num>-sync-testing'.  
✅The pull request base branch must be a 'main' or 'preview' branch..  
✅The pull request head branch contains a valid issue number.  
✅The pull request title matches the linked issue title exactly.  
✅The pull request assignees match the assignees of the issue.  
✅The pull request labels match the labels of the issue.  
✅The pull request organizational projects match the organizational projects of the issue.  
✅The pull requests milestone matches the milestone of the issue.  

---

### **Additional Info:**
1. Pull requests are automatically synced with the associated issue upon creation by one of the KinsonDigital workflow bots. 
2. The associated issue is the issue number that is embedded in the pull request head branch.
3. The list above will be automatically updated as the pull request's various settings match or do not match the associated issue.
4. To manually sync the pull request to the issue, create a comment with the `[run-sync]` command.
	> **Note** You must be an admin member of the organization to use this command.

---

- [x] Sync with the issue.  Use this to enable or disable this pull request from syncing with its associated issue.

### CODE REVIEW

**Review by KinsonDigitalAdmin — DISMISSED** [review]

File: Carbonate/NonDirectional/PushReactable.cs [hunk 1]
```diff
@@ -54,7 +54,7 @@ public void Push(Guid eventId)
     /// <summary>
     /// Sends an error to all of the subscribers that matches the given <paramref name="eventId"/>.
```
Comment (kselena) [inline Carbonate/NonDirectional/PushReactable.cs]: Grammar
```suggestion
    /// Sends an error to all of the subscribers that match the given <paramref name="eventId"/>.
```

File: Carbonate/UniDirectional/PushReactable.cs [hunk 2]
```diff
@@ -60,7 +60,7 @@ public void Push(in TDataIn data, Guid eventId)
     /// <summary>
     /// Sends an error to all of the subscribers that matches the given <paramref name="eventId"/>.
```
Comment (kselena) [inline Carbonate/UniDirectional/PushReactable.cs]: Refactor to below: 
```suggestion
    /// Sends an error to all of the subscribers that match the given <paramref name="eventId"/>.
```

File: README.md [hunk 3]
```diff
@@ -25,7 +33,131 @@ This library is still under development and is not at v1.0.0 yet!!  However, all
 
 <h2 style="font-weight:bold;border:0" align="center">📖 About Carbonate 📖</h2>
 
-**Carbonate** is a messaging library for various applications, using the observable pattern.  It adds the ability to easily and reliably send messages to different parts and/or systems of an application.
+**Carbonate** is a messaging library that uses the observable pattern.  It adds the ability to easily and reliably push and pull messages to different parts or systems of an application.  This helps promote decoupling between different parts of your application as well as increasing your application's testability.
```
Comment (kselena) [inline README.md]: Reword
```suggestion
**Carbonate** is a messaging library built on the observable pattern, empowering seamless and dependable push-and-pull message handling across various parts or systems within an application. By facilitating that communication, it fosters decoupling among different components, enhancing your application's overall testability.
```

File: README.md [hunk 4]
```diff
@@ -25,7 +33,131 @@ This library is still under development and is not at v1.0.0 yet!!  However, all
...
+
+> **Note** Click [here](https://github.com/KinsonDigital/Carbonate/tree/preview/Samples) to view all of the sample projects.
+
+<h2 style="font-weight:bold;border:0" align="center">✨ Features & Benefits ✨</h2>
+
+**Features:**
+- Sending notifications of events with no data
+- Sending notifications of events with data
+- Sending notifications of events with data that also returns data
+- Interfaces and abstractions are provided for custom implementations and to accommodate testing
```
Comment (kselena) [inline README.md]: Reword
```suggestion
- Sends notifications of events with no data
- Sends notifications of events with data
- Sends notifications of events with data and returns data
- Interfaces and abstractions are provided for custom implementations and accommodates testing
```

File: README.md [hunk 5]
```diff
@@ -25,7 +33,131 @@ This library is still under development and is not at v1.0.0 yet!!  However, all
 
 <h2 style="font-weight:bold;border:0" align="center">📖 About Carbonate 📖</h2>
 
-**Carbonate** is a messaging library for various applications, using the observable pattern.  It adds the ability to easily and reliably send messages to different parts and/or systems of an application.
+**Carbonate** is a messaging library that uses the observable pattern.  It adds the ability to easily and reliably push and pull messages to different parts or systems of an application.  This helps promote decoupling between different parts of your application as well as increasing your application's testability.
+
+For a real-world example, check out the [Velaptor](https://github.com/KinsonDigital/Velaptor) code base which is an open source 2D game development framework.  This library has been vital for decoupling the different sub-systems and making Velaptor testable.
+
+Go [here](https://refactoring.guru/design-patterns/observer) for a great explanation of the observer pattern.  There are many tutorials and examples of this pattern all over the web.  It is well documented, well known, and a highly used programming pattern.
```
Comment (kselena) [inline README.md]: Reword
```suggestion
Go [here](https://refactoring.guru/design-patterns/observer) for information on the observer pattern. This design pattern has been extensively covered in various tutorials and examples across the web, making it well-documented, widely recognized, and a highly popular programming pattern.
```

File: README.md [hunk 6]
```diff
@@ -25,7 +33,131 @@ This library is still under development and is not at v1.0.0 yet!!  However, all
...
+**Features:**
+- Sending notifications of events with no data
+- Sending notifications of events with data
+- Sending notifications of events with data that also returns data
+- Interfaces and abstractions are provided for custom implementations and to accommodate testing
+
+**Benefits:**
+- Increases decoupling
+- Increases testability
+- Ability to send data and events without the need to change the public API
```
Comment (kselena) [inline README.md]: Reword
```suggestion
- Sends data and events without changing the public API
```

File: README.md [hunk 7]
```diff
@@ -25,7 +33,131 @@ This library is still under development and is not at v1.0.0 yet!!  However, all
...
+- Promotes the [Open/Closed Principle](https://www.tutorialsteacher.com/csharp/open-closed-principle)
+
+<h2 style="font-weight:bold;border:0" align="center">💡 Examples 💡</h2>
+
+Below are some examples to demonstrate some basic uses of ***Carbonate***.  Use cases will depend on your application and your needs.  ***Carbonate*** is very flexible but how you use it depends on your needs.
```
Comment (kselena) [inline README.md]: Reword
```suggestion
Below are some examples to demonstrate some basic uses of ***Carbonate***.  This library is very flexible but how you use it depends on the needs of your application.
```

File: README.md [hunk 8]
```diff
@@ -25,7 +33,131 @@ This library is still under development and is not at v1.0.0 yet!!  However, all
...
+
+<h3 style="font-weight:bold;color: #00BBC6">Non-directional push notifications</h3>
+
+To send a _**non-directional**_ push notification of something that has occurred, you can use the `PushReactable` type. The subscriber will use the `ReceiveReactor` type for subscriptions. The term _**Non-directional**_ means that no data is being pushed out with the notification or returned from the notification.  This is great for sending a notification that an event has occurred when no data is needed.
+
```
Comment (kselena) [inline README.md]: Reword
```suggestion
To send a _**non-directional**_ push notification of something that has occurred, you can use the `PushReactable` type. The subscriber will use the `ReceiveReactor` type for subscriptions. The term _**Non-directional**_ means that no data is being pushed out or returned from the notification.  This is great for sending a notification that an event has occurred when no data is needed.

```

File: README.md [hunk 9]
```diff
@@ -25,7 +33,131 @@ This library is still under development and is not at v1.0.0 yet!!  However, all
...
+    onReceive: () => Console.WriteLine("Received a message!"),
+    onUnsubscribe: () => Console.WriteLine("Unsubscribed from notifications!"),
+    onError: (ex) => Console.WriteLine($"Error: {ex.Message}")
+));
+
+messenger.Push(msgEventId); // Will invoke all onReceive 'Actions'
+unsubscriber.Dispose(); // Will only unsubscribe from this subscription
+```
+
+Each notification sent out has a unique id.  Any subscribers to that notification must use the same unique ID to receive that notification.  This is to make sure that the notification is unique and that no logic is required with the other subscriptions to make sure that the notification was meant for them.
```
Comment (kselena) [inline README.md]: Refactor "id" to "ID".
```suggestion
Every notification sent out contains a unique ID, which subscribers must use to receive the intended notification, ensuring its exclusivity and eliminating the need for additional logic with other subscriptions.
```

File: README.md [hunk 10]
```diff
@@ -25,7 +33,131 @@ This library is still under development and is not at v1.0.0 yet!!  However, all
...
+
+> **💡TIP💡**  
+> If you want to only receive a single notification, just unsubscribe from further notifications by calling the `Dispose()`
```
Comment (kselena) [inline README.md]: Reword
```suggestion
> If you want to receive a single notification, unsubscribe from further notifications by calling the `Dispose()`
```

File: README.md [hunk 11]
```diff
@@ -25,7 +33,131 @@ This library is still under development and is not at v1.0.0 yet!!  However, all
...
+
+> **Some notes about exceptions and unsubscribing**
+> - Throwing an exception in the 'onReceive' action implementation will invoke the 'onError' action for _**ALL**_ subscriptions.
+> - Invoking `messenger.Dispose()` will invoke the `onUnsubscribe` action for _**ALL**_ subscriptions.
+> - You can unsubscribe from a single subscription by calling the `Dispose()` method on the `IDisposable` object returned by the reactable's `Subscribe()` method.
+
+
+<h3 style="font-weight:bold;color: #00BBC6">Uni-directional push notifications</h3>
+
+To send data with a push notification in one direction, you can use the `PushReactable<T>` type.  The subscriber will use the `ReceiveReactor<T>` type for subscriptions.  This is set up and used just like the previous example. The term _**one-directional**_ means that data travels in one direction which is out with the push notification.
```
Comment (kselena) [inline README.md]: Reword
```suggestion
To facilitate one-directional data transfer through push notifications, you can employ the PushReactable<T> type for sending data, while subscribers utilize the ReceiveReactor<T> type for their subscriptions. Setting up and using this approach follows the same steps as in the previous example. In this context, the term one-directional signifies that data exclusively flows outward with the push notification.
```

File: README.md [hunk 12]
```diff
@@ -25,7 +33,131 @@ This library is still under development and is not at v1.0.0 yet!!  However, all
...
+));
+
+messenger.Push("hello world!", msgEventId); // Will invoke all onReceive 'Actions'
+messenger.Unsubscribe(msgEventId); // Will invoke all onUnsubscribe 'Actions'
+```
+
+
+<h3 style="font-weight:bold;color: #00BBC6">Bi-directional push notifications</h3>
+
+To send a _**bi-directional**_ push notification that sends data out and gets data back, you can use the `PullReactable<T, T>` type.  The subscriber will use the `RespondReactor<T, T>` to subscribe to notifications.  This is used when you want to send a push notification with data that the receiver needs, and then the receiver returns data to the caller that pushed the notification.
```
Comment (kselena) [inline README.md]: Reword
```suggestion
To enable bi-directional push notifications, allowing data to be sent out and received back, you can employ the PullReactable<T, T> type. Subscribers, on the other hand, utilize the RespondReactor<T, T> for their notification subscriptions. This approach proves useful when you need to send a push notification with data required by the receiver, who then responds with data back to the original caller that initiated the notification.
```

File: README.md [hunk 13]
```diff
@@ -25,7 +33,131 @@ This library is still under development and is not at v1.0.0 yet!!  However, all
...
+    onError: (ex) => Console.WriteLine($"Error: {ex.Message}")
+));
+
+Console.WriteLine($"Favorite Language: {favoriteGetter.Pull("prog-lang", msgEventId)}");
+Console.WriteLine($"Favorite Food: {favoriteGetter.Pull("food", msgEventId)}");
+Console.WriteLine($"Favorite Past Time: {favoriteGetter.Pull("past-time", msgEventId)}");
+Console.WriteLine($"Favorite Music: {favoriteGetter.Pull("music", msgEventId)}");
+```
+
+> **Note** The difference between _**bi-directional**_ and _**uni-directional**_ notifications is that _**bi-directional**_ notifications send data out and expected data to be returned.
```
Comment (kselena) [inline README.md]: Reword.  
```suggestion
> **Note** The difference between _**bi-directional**_ and _**uni-directional**_ notifications is that _**bi-directional**_ notifications enable data exchange in both directions whereas _**uni-directional**_ notifications send data out and do not expect data to be returned. 
```

File: README.md [hunk 14]
```diff
@@ -25,7 +33,131 @@ This library is still under development and is not at v1.0.0 yet!!  However, all
...
+
+> **💡TIP💡** Most of the time, the `PushReactable`, `PushReactable<T>`, and `PullReactable<string, string>` types will suit your needs.  But if you have any requirements that these cannot provide, you can always create implementations of the interfaces provided.
```
Comment (kselena) [inline README.md]: Reword
```suggestion
> **💡TIP💡** Most of the time, the `PushReactable`, `PushReactable<T>`, and `PullReactable<string, string>` types will suit your needs.  However, if you have requirements that these can't provide, you can create implementations of the interfaces provided.
```

**Review by kselena — CHANGES_REQUESTED** [review]

**Review by kselena — APPROVED** [review]

**Review by KinsonDigitalAdmin — APPROVED** [review]

**Pull request merged**